File: backend/crates/domain/delta-usecase/src/interactor/claude_format.rs
//! Claude Code text-format detection, in one place.
//!
//! These are the textual conventions Claude Code uses on the wire Delta
//! observes (the JSONL transcript and the hook payloads), gathered here so the
//! transcript→input conversion is the only layer that knows them. The
//! structural flags (e.g. `is_queued_command`) are already detected by the
//! transcript parser in the gateway; these cover the conventions that are
//! plain strings.

/// Prefix Claude Code writes to the transcript when the user interrupts the
/// in-flight turn. It appears as a `role: user` line whose only text block is
/// either `[Request interrupted by user]` (plain mid-response interrupt) or
/// `[Request interrupted by user for tool use]` (interrupt during a tool use).
/// Matching on the shared prefix covers both variants (and any future suffix)
/// without enumerating each exact string.
const INTERRUPT_MARKER_PREFIX: &str = "[Request interrupted by user";

/// Prompt prefix Claude Code uses when it injects a background-task
/// completion notification. Such a submission is a harness injection, not a
/// human typing into the pane, so it must not be reported as external input.
const TASK_NOTIFICATION_PREFIX: &str = "<task-notification>";

/// Whether a (trimmed) user-line text is the interrupt marker Claude Code
/// writes when the user aborts the in-flight turn.
pub(in crate::interactor) fn is_interrupt_marker(trimmed_text: &str) -> bool {
    trimmed_text.starts_with(INTERRUPT_MARKER_PREFIX)
}

/// Whether a hook-submitted prompt is a harness-injected task notification
/// rather than something typed into the pane.
pub(in crate::interactor) fn is_task_notification(prompt: &str) -> bool {
    prompt.trim_start().starts_with(TASK_NOTIFICATION_PREFIX)
}

File: frontend/packages/apps/web/src/features/transcript/claudeFormat.ts
import type { Message } from '@delta/wire-gen';

/**
 * Claude Code text-format detection for the frontend, in one place.
 *
 * Mirrors the backend conventions (see the `delta-attribution` crate's
 * `claude_format` module). The backend owns role/attribution; the frontend
 * only needs these prefixes for presentational decisions, so the detection is
 * duplicated here verbatim rather than shipped over the wire.
 */

/**
 * Prompt prefix Claude Code uses when it injects a background-task completion
 * notification. Such a submission is a harness injection the agent responds to,
 * not human-authored prose, so the conversation pane folds it by default.
 */
const TASK_NOTIFICATION_PREFIX = '<task-notification>';

/**
 * Whether a (trimmed) prompt string is a harness-injected task notification.
 * Matches the backend `is_task_notification`: trim leading whitespace, then
 * check the prefix.
 */
export function isTaskNotificationText(text: string): boolean {
  return text.trimStart().startsWith(TASK_NOTIFICATION_PREFIX);
}

/**
 * Whether a message is a task-notification user turn. The harness submits the
 * notification as a normal `role: "user"` line (not a meta row), so the
 * meta-folding does not apply; the conversation pane detects it here and folds
 * the block presentationally. Backend role/attribution is unchanged.
 *
 * The notification text arrives as the first `text` block of the message, so
 * the check mirrors the backend prompt-prefix detection on that block.
 */
export function isTaskNotificationMessage(message: Message): boolean {
  if (message.role !== 'user') {
    return false;
  }
  const firstText = message.content.find((block) => block.type === 'text');
  return firstText !== undefined && isTaskNotificationText(firstText.text);
}
